### DISCUSSION TITLE
esp32_ULP hiding place

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#12688] [Libraries & Drivers] [answered]
I last worked with the esp32s ULP back at about upython v17 &, although I found the ULP hard work, getting the necessaries in place was easy
```
import upip; upip.install('micropython-py-esp32-ulp') 
```
I've tried to move with the times now that I'm using V21 but it hasn't worked
```
>>> import mip; mip.install('micropython-py-esp32-ulp')
Installing micropython-py-esp32-ulp (latest) from https://micropython.org/pi/v2 to /lib
Package not found: https://micropython.org/pi/v2/package/6/micropython-py-esp32-ulp/latest.json
Package may be partially installed
```
There is no lib file & the import
```
from esp32_ulp import src_to_binary
ImportError: no module named 'esp32_ulp'
```
 crashes. I guess I could roll back to v17 but I prefer to stick with v21. Does anybody know how to get the esp32_ulp folder onto a board running v21?

**Comment by steveh127** [comment]
I think this is what you want: [https://github.com/micropython/micropython-esp32-ulp](url), there are install instructions for v1.20+ which worked for me on an ESP32 S3 board running the latest 1.22 preview.

**Comment by kjm1102** [comment]
That link gives me a 404 steve but I eventually found directions to the esp32_ulp folder @ https://github.com/micropython/micropython-esp32-ulp (mip.install('github:micropython/micropython-esp32-ulp'))

**Reply by steveh127** [reply, marked as answer]
Not sure what happened there. I thought I'd copied the link OK, it's now corrected. Perhaps mark as answered?
